packages/remix-dev/vite/plugin.ts

```diff
@@ -1527,34 +1527,46 @@ async function handleSpaMode(
   let { createRequestHandler: createHandler } = await import("@remix-run/node");
   let handler = createHandler(build, viteConfig.mode);
   let response = await handler(new Request("http://localhost/"));
-  let html = await response.text();
+  let text = await response.text();
+
   if (response.status !== 200) {
     throw new Error(
       `SPA Mode: Received a ${response.status} status code from ` +
-        `\`entry.server.tsx\` while generating the \`index.html\` file.\n${html}`
+        `\`entry.server.tsx\` while generating the \`index.html\` file.\n${text}`
     );
   }
 
-  if (
-    !html.includes("window.__remixContext =") ||
-    !html.includes("window.__remixRouteModules =")
+  if (response.headers.get("Content-Type") === "application/javascript") {
+    let jsDir = path.join(assetsBuildDirectory, "assets");
+    let jsPath = path.join(jsDir, "index.js");
+    await fse.writeFile(jsPath, text);
+
+    viteConfig.logger.info(
+      "SPA Mode: index.js has been written to your " +
+        colors.bold(path.relative(process.cwd(), jsDir)) +
+        " directory"
+    );
```

Comment:
If the user returns a `application/javascript` response from the handler, we write it out as JS instead of HTML